enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,612 @@
+    - CRI-O:
+      - `assets/performanceprofile/configs/99-workload-pinning.conf`
+      - `assets/performanceprofile/configs/99-runtimes.conf`: `infra_ctr_cpuset = "{{.ReservedCpus}}"`
+    - Sysctl: `assets/performanceprofile/configs/99-default-rps-mask.conf`: RPS Mask == .Reserved
+    - Kubernetes: `assets/performanceprofile/configs/openshift-workload-pinning`
+    - Tuned: `/sys/devices/system/cpu/cpufreq/policy{{ <<CPU>> }}/scaling_max_freq={{$.ReservedCpuMaxFreq}}`
+
+  - .Isolated - CPU set for the application container workloads. Should be used for low latency workload.
+    > Relevant for low latency
+    - Tuned: `isolated_cores={{.IsolatedCpus}}`
+    - Tuned: `/sys/devices/system/cpu/cpufreq/policy{{ <<CPU>> }}/scaling_max_freq={{$.IsolatedCpuMaxFreq}}`
+      > Impossible to do without dynamic templating (each CPU in .Isolated CPU set needs separate line)
+
+  - .BalanceIsolated - toggles if the .Isolated CPU set is eligible for load balancing work loads.
+     If `false`, Isolated CPU set is static, meaning workloads have to explicitly assign each thread
+     to a specific CPU in order to work across multiple CPUs.
+     - Tuned: true -> cmdline isolcpus=domain,managed_irq,${isolated_cores}, otherwise isolcpus=managed_irq,${isolated_cores}
+       > Not implemented. Users can use `cpu-load-balancing.crio.io` annotation instead.
+
+  - .Offlined - CPU set be unused and set offline.
+    > Out of scope
+    - Systemd: unit running `assets/performanceprofile/scripts/set-cpus-offline.sh`
+
+  - .Shared - CPU set shared among guaranteed workloads needing additional CPUs which are not exclusive.
+    > User configures in Kubelet config
+    - KubeletConfig: if .MixedCPUsEnabled=true, then .ReservedSystemCPUs = .Reserved Union .Shared
+
+- .HardwareTuning
+  - Tuned: if !.PerPodPowerManagement, then cmdline =+ `intel_pstate=active`
+    > cpu-partitioning sets `intel_pstate=disable`, if user wants different value they can use
+    > `additionalArgs` in `microshift-low-latency-variables.conf` - in case of duplicated parameters,
+    > last one takes precedence
+  - .IsolatedCpuFreq (int) - defines a minimum frequency to be set across isolated cpus
+    - Tuned: `/sys/devices/system/cpu/cpufreq/policy{{ <<CPU>> }}/scaling_max_freq={{$.IsolatedCpuMaxFreq}}`
+      > Not doable without dynamic templating
+  - .ReservedCpuFreq (int) - defines a minimum frequency to be set across reserved cpus
+    - Tuned: `/sys/devices/system/cpu/cpufreq/policy{{ <<CPU>> }}/scaling_max_freq={{$.ReservedCpuMaxFreq}}`
+      > Not doable without dynamic templating
+
+- .HugePages
+  - .DefaultHugePagesSize (string)
+    - Tuned: cmdline =+ default_hugepagesz=%s
+      > Handled
+  - .Pages (slice)
+    - .Size
+      - Tuned: cmdline =+ hugepagesz=%s
+        > Handled
+    - .Count
+      - Tuned: cmdline =+ hugepages=%d
+        > Handled
+    - .Node - NUMA node, if not provided, hugepages are set in kernel args
```

Comment:
The user could alternatively use the [tuned sysfs plugin](https://github.com/redhat-performance/tuned/blob/master/doc/manual/modules/performance/ref_available-tuned-plug-ins.adoc), 512 2M hugepages to numa node0:
```
[sysfs]
 /sys/devices/system/node/node0/hugepages/hugepages-2048kB/nr_hugepages = 512
```

Reply:
And this is another problem, if we ship tuned profile, then any user customization requires them to create a new tuned profile, unless we take a page out of NTO's book and do some dynamic generation. Right?

Reply:
We could either create variables that the user can set the values themselves in `/etc/tuned/microshift-low-latency-variables.conf` or just have the user create their own tuned profile that will include the `microshift-low-latency`. I'm not suggesting anything dynamic at this point.